### PR TITLE
refactor: update to typescript definitions from juggler

### PR DESCRIPTION
**Pull request by raymondfeng**
This PR upgrades to `loopback-datasource-juggler@^3.20.0` to leverage TypeScript definitions which are now shipped with the juggler module.

## Checklist

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by raymondfeng** [comment]
> I might be forgetting some other files but make sure tweak docs.json in repository so that it doesn't include loopback-datasource-juggler anymore

Good catch. Removed.

**Comment by shimks** [comment]
Was the node 10 commit accidentally pushed to this branch? I think it makes more sense to leave the upgrades to a separate PR

File: packages/repository/src/common-types.ts [hunk 1]
```diff
@@ -51,7 +51,7 @@ export type DataObject<T> = T | AnyObject;
 /**
  * Type alias for Node.js options object
  */
-export type Options = AnyObject | null | undefined;
+export type Options = AnyObject | undefined;
```
Comment (virkt25) [inline packages/repository/src/common-types.ts]: Why is `null` no longer needed here?

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 2]
```diff
@@ -37,9 +34,9 @@ export const ModelBaseConstructor = jugglerModule.ModelBaseClass as typeof juggl
  * @param ds {DataSource} Data source
  * @returns {} The new model class with DAO (CRUD) operations
  */
-export function bindModel<T extends typeof juggler.ModelBase>(
+export function bindModel<T extends legacy.ModelBaseClass>(
   modelClass: T,
-  ds: juggler.DataSource,
+  ds: legacy.DataSource,
```
Comment (virkt25) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: Is it possible to use `juggler.DataSource` here -- the namespace being exported above?

**Comment by raymondfeng** [comment]
> Was the node 10 commit accidentally pushed to this branch? I think it makes more sense to leave the upgrades to a separate PR.

I was testing a fix. See https://github.com/strongloop/loopback-next/pull/1321

File: packages/repository/src/common-types.ts (hunk 1 above)
Comment (raymondfeng) [reply packages/repository/src/common-types.ts]: See https://basarat.gitbooks.io/typescript/docs/javascript/null-undefined.html

Comment (raymondfeng) [reply packages/repository/src/common-types.ts]: We should even remove `undefined` from `Options`. It should be `options?: Options` instead.

File: packages/repository/src/mixins/repository.mixin.ts [hunk 3]
```diff
@@ -38,7 +38,7 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
      *   model: any;
      *
      *   constructor() {
-     *     const ds: juggler.DataSource = new DataSourceConstructor({
+     *     const ds: juggler.DataSource = new juggler.DataSource({
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/repository/src/mixins/repository.mixin.ts]: I did that for purpose to avoid confusions between lb4 and lb3 classes.

**Pull request merged**